cmd/bosun/web/middlewares.go

```diff
@@ -0,0 +1,132 @@
+package web
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/MiniProfiler/go/miniprofiler"
+	"github.com/captncraig/easyauth"
+	"github.com/captncraig/easyauth/providers/ldap"
+	"github.com/captncraig/easyauth/providers/token"
+	"github.com/captncraig/easyauth/providers/token/redisStore"
+	"github.com/gorilla/mux"
+
+	"bosun.org/cmd/bosun/conf"
+	"bosun.org/collect"
+	"bosun.org/metadata"
+	"bosun.org/opentsdb"
+)
+
+// custom middlewares for bosun. Must match  alice.Constructor signature (func(http.Handler) http.Handler)
+
+var miniprofilerMiddleware = func(next http.Handler) http.Handler {
+	return miniprofiler.NewContextHandler(next.ServeHTTP)
+}
+
+var endpointStatsMiddleware = func(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		//metric for http vs https
+		proto := "http"
+		if r.TLS != nil {
+			proto = "https"
+		}
+		collect.Add("http_protocol", opentsdb.TagSet{"proto": proto}, 1)
+
+		//if we use gorilla named routes, we can add stats and timings per route
+		routeName := ""
+		if route := mux.CurrentRoute(r); route != nil {
+			routeName = route.GetName()
+		}
+		if routeName == "" {
+			routeName = "unknown"
+		}
+		t := collect.StartTimer("http_routes", opentsdb.TagSet{"route": routeName})
+		next.ServeHTTP(w, r)
+		t()
+	})
+}
+
+type noopAuth struct{}
+
+func (n noopAuth) GetUser(r *http.Request) (*easyauth.User, error) {
+	//TODO: make sure ui sends header when possible, instead of json body
+	name := "anonymous"
+	if cookie, err := r.Cookie("action-user"); err == nil {
+		name = cookie.Value
+	}
+	//everybody is an admin!
+	return &easyauth.User{
+		Access:   roleAdmin,
+		Username: name,
+		Method:   "noop",
+	}, nil
+}
+
+func buildAuth(cfg conf.AuthConf) (easyauth.AuthManager, *token.TokenProvider, error) {
+	if cfg.CookieSecret == "" {
+		cfg.CookieSecret = "CookiesAreInsecure"
+	}
+	auth, err := easyauth.New(easyauth.CookieSecret(cfg.CookieSecret))
+	if err != nil {
+		return nil, nil, err
+	}
+	if cfg.AuthEnabled == false {
+		auth.AddProvider("nop", noopAuth{})
+	} else {
+		authEnabled = true
+		// in proc token store so bosun can send data to itself with an ephemeral token
+		selfStore, _ := token.NewJsonStore("")
+		selfToken := token.NewToken(easyauth.RandomString(16), selfStore)
```

Comment:
16 bytes is an adequate hash secret. magic number because its not worth making a constant over.